### PR TITLE
Port DistanceMachine.cpp to use OpenMP

### PR DESCRIPTION
**Pull request by abhinavrai44**

### CODE REVIEW

**Comment by vigsterkr** [comment]
@abhinavrai44 could you rebase with the latest develop?

File: src/shogun/distance/Distance.cpp [hunk 1]
```diff
@@ -216,6 +216,18 @@ float64_t CDistance::distance(int32_t idx_a, int32_t idx_b)
 	return compute(idx_a, idx_b);
 }
 
+float64_t* CDistance::run_distance_rhs(float64_t* result, int32_t idx_r_start, int32_t idx_start, int32_t idx_stop, int32_t idx_a)
+{
+	for(int32_t i=idx_r_start; idx_start < idx_stop; i++,idx_start++)
+        result[i] = this->distance(idx_a,idx_start);
```
Comment (vigsterkr) [inline src/shogun/distance/Distance.cpp]: no return value, hence CI is failing

File: src/shogun/distance/Distance.cpp [hunk 2]
```diff
@@ -216,6 +216,18 @@ float64_t CDistance::distance(int32_t idx_a, int32_t idx_b)
...
+}
+
+float64_t* CDistance::run_distance_lhs(float64_t* result, int32_t idx_r_start, int32_t idx_start, int32_t idx_stop, int32_t idx_b)
+{
+	for(int32_t i=idx_r_start; idx_start < idx_stop; i++,idx_start++)
+        result[i] = this->distance(idx_start,idx_b);
```
Comment (vigsterkr) [inline src/shogun/distance/Distance.cpp]: return value is missing...

File: src/shogun/distance/Distance.h [hunk 3]
```diff
@@ -335,6 +335,28 @@ class CDistance : public CSGObject
 			return lhs==rhs;
 		}
 
+		/**
```
Comment (vigsterkr) [inline src/shogun/distance/Distance.h]: please write more description in the doxygen comments, especially about the parameters...

File: src/shogun/distance/Distance.h [hunk 4]
```diff
@@ -335,6 +335,28 @@ class CDistance : public CSGObject
...
+		 * Function for computing distance values
+		 *
+		 * @param result array of distance values
+		 * @param idx_r_start 
+		 * @param idx_start
+		 * @param idx_stop
+		 * @param idx_a feature vector a at idx_a
+		 */
+		float64_t* run_distance_rhs(float64_t* result, int32_t idx_r_start, int32_t idx_start, int32_t idx_stop, int32_t idx_a);
```
Comment (vigsterkr) [inline src/shogun/distance/Distance.h]: why do you pass native `float64_t*` arrays, when we have `SGVector<float64_t>` for such things...

Comment (vigsterkr) [inline src/shogun/distance/Distance.h]: and please specify whether an input argument should be `const` or not... like most of the integers could be...

File: src/shogun/distance/Distance.h [hunk 5]
```diff
@@ -335,6 +335,28 @@ class CDistance : public CSGObject
...
+		/**
+		 * Function for computing distance values
+		 *
+		 * @param result array of distance values
+		 * @param idx_r_start
+		 * @param idx_start
+		 * @param idx_stop
+		 * @param idx_b feature vector b at idx_b
+		 */
+		float64_t* run_distance_lhs(float64_t* result, int32_t idx_r_start, int32_t idx_start, int32_t idx_stop, int32_t idx_b);
```
Comment (vigsterkr) [inline src/shogun/distance/Distance.h]: same as above

File: src/shogun/machine/DistanceMachine.cpp [hunk 6]
```diff
@@ -51,168 +43,68 @@ void CDistanceMachine::init()
...
 
-        pthread_attr_destroy(&attr);
-        SG_FREE(params);
-        SG_FREE(threads);
-    }
+        distance->run_distance_lhs(result, idx_r_start, idx_start, idx_stop, idx_b);
+#else
+        int32_t idx_r_start = idx_a1;
+        int32_t idx_start = idx_a1;
+        int32_t idx_stop = idx_a2  + 1;
```
Comment (vigsterkr) [inline src/shogun/machine/DistanceMachine.cpp]: extra whitespace

File: src/shogun/machine/DistanceMachine.cpp [hunk 7]
```diff
@@ -51,168 +43,68 @@ void CDistanceMachine::init()
...
-
-        pthread_attr_destroy(&attr);
-        SG_FREE(params);
-        SG_FREE(threads);
-    }
+        distance->run_distance_rhs(result, idx_r_start, idx_start, idx_stop, idx_a);
+#else
+        int32_t idx_r_start = idx_b1;
+        int32_t idx_start = idx_b1;
+        int32_t idx_stop = idx_b2  + 1;
```
Comment (vigsterkr) [inline src/shogun/machine/DistanceMachine.cpp]: extra whitespace

**Comment by vigsterkr** [comment]
the comments regarding the arrays being used natively instead of SGVector has not been addressed by your fix.

one more thing:
please used `index_t` type for indecies like for example:
```
       int32_t idx_r_start = idx_a1;
       int32_t idx_start = idx_a1;
       int32_t idx_stop = idx_a2  + 1; 
```

**Comment by abhinavrai44** [comment]
@vigsterkr Plz review

**Comment by abhinavrai44** [comment]
@vigsterkr Plz review the changes

File: src/shogun/distance/Distance.cpp [hunk 8]
```diff
@@ -216,6 +216,18 @@ float64_t CDistance::distance(int32_t idx_a, int32_t idx_b)
 	return compute(idx_a, idx_b);
 }
 
+void CDistance::run_distance_rhs(SGVector<float64_t>& result, const index_t idx_r_start, index_t idx_start, const index_t idx_stop, const int32_t idx_a)
+{
+	for(int32_t i=idx_r_start; idx_start < idx_stop; i++,idx_start++)
```
Comment (vigsterkr) [inline src/shogun/distance/Distance.cpp]: this should be `index_t` just as any indexing variable....
you can do `++i` and `++idx_start`

File: src/shogun/distance/Distance.cpp [hunk 9]
```diff
@@ -216,6 +216,18 @@ float64_t CDistance::distance(int32_t idx_a, int32_t idx_b)
...
+        result.vector[i] = this->distance(idx_a,idx_start);
+}
+
+void CDistance::run_distance_lhs(SGVector<float64_t>& result, const index_t idx_r_start, index_t idx_start, const index_t idx_stop, const int32_t idx_b)
+{
+	for(int32_t i=idx_r_start; idx_start < idx_stop; i++,idx_start++)
```
Comment (vigsterkr) [inline src/shogun/distance/Distance.cpp]: this should be `index_t` just as any indexing variable....
you can do `++i` and `++idx_start`

File: src/shogun/distance/Distance.cpp [hunk 10]
```diff
@@ -216,6 +216,18 @@ float64_t CDistance::distance(int32_t idx_a, int32_t idx_b)
 	return compute(idx_a, idx_b);
 }
 
+void CDistance::run_distance_rhs(SGVector<float64_t>& result, const index_t idx_r_start, index_t idx_start, const index_t idx_stop, const int32_t idx_a)
```
Comment (vigsterkr) [inline src/shogun/distance/Distance.cpp]: all indices should be `index_t`

File: src/shogun/distance/Distance.cpp [hunk 11]
```diff
@@ -216,6 +216,18 @@ float64_t CDistance::distance(int32_t idx_a, int32_t idx_b)
...
+{
+	for(int32_t i=idx_r_start; idx_start < idx_stop; i++,idx_start++)
+        result.vector[i] = this->distance(idx_a,idx_start);
+}
+
+void CDistance::run_distance_lhs(SGVector<float64_t>& result, const index_t idx_r_start, index_t idx_start, const index_t idx_stop, const int32_t idx_b)
```
Comment (vigsterkr) [inline src/shogun/distance/Distance.cpp]: all indices should be `index_t`

File: src/shogun/machine/DistanceMachine.cpp [hunk 12]
```diff
@@ -49,170 +41,68 @@ void CDistanceMachine::init()
 	m_parameters->add((CSGObject**)&distance, "distance", "Distance to use");
 }
 
-void CDistanceMachine::distances_lhs(float64_t* result,int32_t idx_a1,int32_t idx_a2,int32_t idx_b)
+void CDistanceMachine::distances_lhs(SGVector<float64_t>& result, int32_t idx_a1, int32_t idx_a2, int32_t idx_b)
```
Comment (vigsterkr) [inline src/shogun/machine/DistanceMachine.cpp]: if it's and index it's an `index_t`

File: src/shogun/distance/Distance.cpp (hunk 10 above)
Comment (abhinavrai44) [inline src/shogun/distance/Distance.cpp]: Sorry. My mistake.

**Comment by abhinavrai44** [comment]
@vigsterkr Done

**Comment by vigsterkr** [comment]
yey great thnx!

**Comment by abhinavrai44** [comment]
Thanks

**Pull request merged**